src/util.cpp

```diff
@@ -375,6 +376,26 @@ int LogPrintStr(const std::string &str)
     return ret;
 }
 
+bool LockDirectory(const fs::path& directory, const std::string lockfile_name, bool probe_only)
+{
+    fs::path pathLockFile = directory / lockfile_name;
+    FILE* file = fsbridge::fopen(pathLockFile, "a"); // empty lock file; created if it doesn't exist.
+    if (file) fclose(file);
+
+    try {
+        static boost::interprocess::file_lock lock(pathLockFile.string().c_str());
```

Comment:
From https://botbot.me/freenode/bitcoin-core-dev/msg/95065183/

> I've been trying to work out why the last commit in #11904 is breaking the locking function, it works fine before that commit is added, but I can't work out why it doesn't work with that change. with a brand new walletdir, it creates the .walletlock file, but lock.try_lock() fails and the node crashes on startup, does anyone with more boost familiarity than me know why lock.try_lock fails when I move it

I haven't looked closely at changes here yet, but as I was trying to say in https://github.com/bitcoin/bitcoin/pull/11904#discussion_r158317339, I think the static declaration on this line becomes broken as soon as more than one directory is locked, because this will only create one single `file_lock` instance no matter how many different directories need to be locked. I think what you really need here is a map or list of locks, not one single lock. Perhaps, something along the lines of (untested):

```c++
static std::map<std::string, boost::interprocess::file_lock> locks;
boost::interprocess::file_lock& lock = locks.emplace(pathLockFile.string(), pathLockFile.string().c_str()).first->second;
```

Reply:
Ah, of course. Fixed. Thank you!